### PR TITLE
Removing proto functionality from REST /predict endpoint [#803]

### LINKED ISSUE #803: High latency of REST locally
I'm trying to [profiling](https://pypi.org/project/profilehooks/) the seldon client locally. I was using `seldon-core-microservice <MyModel> REST` to serve the model. I was using XGBoost model with ~ 30 features and 200 rows. And my script is 
```
@profile
def predict_localhost() -> None:
    sc = SeldonClient()
    p = sc.microservice(
        method="predict",
        payload_type="ndarray",
        data=nparray,
        names=names,
        transport=REST
    )

if __name__ == "__main__":
    for i in range(100):
        predict_localhost()
```

The profiling result shows:
```
*** PROFILER RESULTS ***
predict_localhost (client.py:36)
function called 100 times

         16729181 function calls (14395219 primitive calls) in 10.947 seconds

   Ordered by: cumulative time, internal time, call count
   List reduced from 675 to 40 due to restriction <40>

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
      100    0.000    0.000   10.948    0.109 client.py:36(predict_localhost)
      100    0.001    0.000   10.945    0.109 seldon_client.py:323(microservice)
      100    0.215    0.002   10.943    0.109 seldon_client.py:479(microservice_api_rest_seldon_message)
      100    0.000    0.000    5.884    0.059 api.py:104(post)
      100    0.001    0.000    5.884    0.059 api.py:16(request)
      100    0.001    0.000    5.862    0.059 sessions.py:466(request)
      100    0.002    0.000    4.914    0.049 sessions.py:617(send)
      100    0.001    0.000    4.896    0.049 adapters.py:394(send)
      100    0.002    0.000    4.830    0.048 connectionpool.py:449(urlopen)
      100    0.002    0.000    4.811    0.048 connectionpool.py:320(_make_request)
      100    0.001    0.000    4.686    0.047 client.py:1287(getresponse)
      100    0.002    0.000    4.681    0.047 client.py:290(begin)
      700    0.002    0.000    4.655    0.007 {method 'readline' of '_io.BufferedReader' objects}
      100    0.002    0.000    4.654    0.047 client.py:257(_read_status)
      205    0.001    0.000    4.653    0.023 socket.py:572(readinto)
      205    4.651    0.023    4.651    0.023 {method 'recv_into' of '_socket.socket' objects}
      100    0.001    0.000    3.108    0.031 utils.py:67(seldon_message_to_json)
      100    0.007    0.000    3.046    0.030 json_format.py:99(MessageToJson)
      100    0.001    0.000    3.039    0.030 json_format.py:191(ToJsonString)
  300/100    0.002    0.000    1.709    0.017 json_format.py:195(_MessageToJsonObject)
  200/100    0.002    0.000    1.708    0.017 json_format.py:206(_RegularMessageToJsonObject)
528500/100    0.472    0.000    1.707    0.017 json_format.py:271(_FieldToJsonObject)
20100/100    0.020    0.000    1.702    0.017 json_format.py:349(_ListValueMessageToJsonObject)
20100/100    0.338    0.000    1.701    0.017 json_format.py:351(<listcomp>)
560000/20000    0.511    0.000    1.686    0.000 json_format.py:333(_ValueMessageToJsonObject)
      100    0.271    0.003    1.540    0.015 utils.py:233(array_to_grpc_datadef)
      200    0.007    0.000    1.460    0.007 __init__.py:183(dumps)
      200    0.061    0.000    1.452    0.007 encoder.py:182(encode)
20100/100    0.048    0.000    1.268    0.013 utils.py:279(array_to_list_value)
   604700    0.102    0.000    1.252    0.000 encoder.py:412(_iterencode)
    20000    0.100    0.000    1.168    0.000 well_known_types.py:817(extend)
1208800/604700    0.201    0.000    1.150    0.000 encoder.py:333(_iterencode_dict)
   540000    0.353    0.000    1.069    0.000 well_known_types.py:814(append)
1203300/603300    0.520    0.000    0.949    0.000 encoder.py:277(_iterencode_list)
      100    0.001    0.000    0.778    0.008 sessions.py:426(prepare_request)
      100    0.001    0.000    0.755    0.008 models.py:307(prepare)
      100    0.001    0.000    0.707    0.007 models.py:452(prepare_body)
      200    0.001    0.000    0.704    0.004 models.py:82(_encode_params)
      200    0.001    0.000    0.702    0.004 parse.py:846(urlencode)
      200    0.001    0.000    0.700    0.004 parse.py:807(quote_plus)
```
As you can see the network has a latency of ~ 50 ms (see the api from `connectionpool.py`), and data conversion contributes ~ 30 ms (`MessageToJson` from `protobuf`).  That makes our prediction take ~ 100 + ms, which is too slow.

### PR DESCRIPTION
**Pull request by axsaucedo**
This is the implementation to fix #803. 

Once this is landed there will still be work to replicate across remaining endpoints (as per #607)

### CODE REVIEW

**Comment by lennon310** [comment]
The [profiler result](https://github.com/SeldonIO/seldon-core/issues/803#issuecomment-521844253) shows parsing time reduces for REST but not GRPC, is it expected?

**Comment by axsaucedo** [comment]
@lennon310 that is correct, I've replied in #803, we should be able to merge this soon

**Comment by lennon310** [comment]
Also I was wondering whether this change applies to the prepackaged server as well?

**Comment by axsaucedo** [comment]
@lennon310 yes, although the model servers will also be bumped a version when they are recompiled.

**Comment by axsaucedo** [comment]
Added a couple more tests. PR should be ready for review. @lennon310 could you provide a review as well given that you have been currently going through these pieces?

File: python/seldon_core/utils.py [hunk 1]
```diff
@@ -397,8 +463,12 @@ def construct_response(user_model: SeldonComponent, is_request: bool, client_req
         raise SeldonMicroserviceException("Unknown data type returned as payload:" + client_raw_response)
 
 
-def extract_request_parts_json(request_raw: Union[Dict, List]) -> Tuple[
-    Union[np.ndarray, str, bytes, dict], Dict, prediction_pb2.DefaultData, str]:
+def extract_request_parts_json(request: Union[Dict, List]
```
Comment (lennon310) [inline python/seldon_core/utils.py]: This PR looks good to me. 
Probably not related to this PR, but I was wondering why gRPC path should not go this way. 
From my profiler commented in issue it shows in gRPC the input was first converted to `datadef` in `SeldonMessage`:

```
100    0.240    0.002    1.281    0.013 utils.py:234(array_to_grpc_datadef)
20100/100    0.038    0.000    1.041    0.010 utils.py:280(array_to_list_value)
```

then in `predict` function it is converted back to Numpy array:

```
100    0.000    0.000    2.781    0.028 utils.py:505(extract_request_parts)
100    0.001    0.000    2.778    0.028 utils.py:120(get_data_from_proto)
100    0.001    0.000    2.778    0.028 utils.py:147(grpc_datadef_to_array)
```

The round-trip conversion seems double the latency I guess?

**Comment by axsaucedo** [comment]
@lennon310 the conversion is necessary as in GRPC the request is received as Proto, and needs to be converted into a numpy array before it's passed to the python wrapper. You can access the raw proto through the `predict_raw` function, but if you want to access the data in a readable format you will eventually have to convert it. Similarly when sending the response, it's necessary to convert the output (which is often a numpy array) into a Proto - which in turn requires that conversion.

We're happy to take suggestions but if you have a look at the functions, it's basically converting the Proto directly to numpy array. If you run some alternative implementations of methods to convert numpy arrays to/from proto objects that are more efficient, we would certainly be keen to introduce them.

**Pull request merged**